restli-tools/src/main/java/com/linkedin/restli/tools/data/SchemaFormatTranslator.java

```diff
@@ -207,32 +209,31 @@ private void translateFiles() throws IOException, InterruptedException
   private Map<String, SchemaInfo> getTopLevelSchemaToTranslatedSchemaMap() throws IOException
   {
     Map<String, SchemaInfo> topLevelTranslatedSchemas = new HashMap<>();
+    DataSchemaParser dataSchemaParser = new DataSchemaParser(_resolverPath);
+    DataSchemaParser.ParseResult parsedSources = dataSchemaParser.parseSources(
+        new String[]{_sourceDir.getAbsolutePath()});
 
-    Iterator<File> iter = FileUtils.iterateFiles(_sourceDir, new String[]{_sourceFormat}, true);
-    StringBuilder errorMessages = new StringBuilder();
-    while(iter.hasNext())
+    for(Map.Entry<DataSchema, DataSchemaLocation> entry : parsedSources.getSchemaAndLocations().entrySet())
     {
-      File sourceFile = iter.next();
-      String relativePath = _sourceDir.toURI().relativize(sourceFile.toURI()).getPath();
-      String relativeMinusExt = trimFileExtension(relativePath);
-      String schemaFullname = relativeMinusExt.replace(File.separatorChar, '.');
-
-      // When translating files 1:1, a new resolver and parser are required for each file translated
-      // so that a single top level output schema is matched to each input file.
-      MultiFormatDataSchemaResolver resolver = MultiFormatDataSchemaResolver.withBuiltinFormats(_resolverPath);
-      PegasusSchemaParser parser = AbstractSchemaParser.parserForFileExtension(_sourceFormat, resolver);
-      parser.parse(new FileInputStream(sourceFile));
-      LOGGER.debug("Loaded source schema: {}, from location: {}", schemaFullname, sourceFile.getAbsolutePath());
-      NamedDataSchema schema = checkForErrorsAndGetTopLevelSchema(sourceFile, schemaFullname, parser, errorMessages);
-      if (schema != null)
+      DataSchema schema = entry.getKey();
+      DataSchemaLocation location = entry.getValue();
+      if (!location.getSourceFile().getAbsolutePath().endsWith(_sourceFormat))
       {
-        topLevelTranslatedSchemas.put(schemaFullname, new SchemaInfo(schema, sourceFile, encode(schema)));
+        // Skip source files not matching the format specified.
+        continue;
+      }
+      if (schema instanceof NamedDataSchema)
+      {
+        NamedDataSchema namedDataSchema = (NamedDataSchema) schema;
+        String schemaFullname = namedDataSchema.getFullName();
+        LOGGER.debug("Loaded source schema: {}, from location: {}", schemaFullname, location.getSourceFile().getAbsolutePath());
+        topLevelTranslatedSchemas.put(schemaFullname, new SchemaInfo(namedDataSchema, location.getSourceFile(), encode(schema)));
+      }
+      else
+      {
+        LOGGER.error("Parsed a non-named schema as top-level schema.");
+        System.exit(1);
```

Comment:
Is it worth to just skip the error and keep looping?

Reply:
No, it is a translator. It should have 1 to 1 translation. If there is a non-named schema as a top level schema, it means something wrong. We should throw exception not skip it.

Reply:
This check is mainly a failsafe. The DataSchemaParse::parseSources logic already checks for this and includes errors for all schemas, so this would never be hit.